packages/deployments/contracts/contracts/nomad-xapps/contracts/promise-router/PromiseRouter.sol

```diff
@@ -154,23 +211,58 @@ contract PromiseRouter is Version0, Router {
 
     bytes32 transferId = _msg.transferId();
     address callbackAddress = _msg.callbackAddress();
-    bytes memory data = _msg.returnCallData();
+    bool success = _msg.returnSuccess();
+    bytes memory data = _msg.returnData();
 
-    //TODO process callback
+    //store Promise message
+    promiseMessages[transferId] = _message;
 
     // emit Receive event
-    emit Receive(_originAndNonce(_origin, _nonce), _origin, transferId, callbackAddress, data);
+    emit Receive(_originAndNonce(_origin, _nonce), _origin, transferId, callbackAddress, success, data, _message);
   }
 
-  function process(bytes32 transactionId, bytes memory _message) internal {
-    // Should parse out the return data and callback address from message
-    // Should execute the callback() function on the provided Callback address
-    // Should enforce relayer is whitelisted by calling local connext contract
+  function process(bytes32 transferId) public payable nonReentrant {
```

Comment:
Where function docstring/descriptor? :'((

Reply:
So this method is external now? Basically, we set the `promiseMessages[transferId]` in the `handler`, and anyone can call `process` ?

Reply:
Seems like it should be limited to relayers only, if this is the case. Otherwise we are subject to frontrunning problems

Reply:
also, why payable? lol

Reply:
it checks the relayer in there

Reply:
does not need to be payable though